rig-core/src/embeddings/embeddable.rs

```diff
@@ -0,0 +1,220 @@
+//!     id: "doc1".to_string(),
+//!     word: "hello".to_string(),
+//!     definition: "used as a greeting or to begin a conversation".to_string()
+//! };
+//!
+//! impl Embeddable for FakeDefinition {
+//!     type Error = anyhow::Error;
+//!
+//!     fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error> {
+//!         // Embeddigns only need to be generated for `definition` field.
+//!         // Select it from te struct and return it as a single item.
+//!         Ok(OneOrMany::from(self.definition.clone()))
+//!     }
+//! }
+//! ```
+
+/// Error type used for when the `embeddable` method fails.
+/// Used by default implementations of `Embeddable` for common types.
+#[derive(Debug, thiserror::Error)]
+pub enum EmbeddableError {
+    #[error("SerdeError: {0}")]
+    SerdeError(#[from] serde_json::Error),
+}
+
+/// Trait for types that can be embedded.
+/// The `embeddable` method returns a OneOrMany<String> which contains strings for which embeddings will be generated by the embeddings builder.
+/// If there is an error generating the list of strings, the method should return an error that implements `std::error::Error`.
+pub trait Embeddable {
+    type Error: std::error::Error;
+
+    fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error>;
+}
+
+/// Struct containing either a single item or a list of items of type T.
+/// If a single item is present, `first` will contain it and `rest` will be empty.
+/// If multiple items are present, `first` will contain the first item and `rest` will contain the rest.
+/// IMPORTANT: this struct cannot be created with an empty vector.
+/// OneOrMany objects can only be create using OneOrMany::from() or OneOrMany::try_from().
+#[derive(PartialEq, Eq, Debug)]
+pub struct OneOrMany<T> {
+    /// First item in the list.
+    first: T,
+    /// Rest of the items in the list.
+    rest: Vec<T>,
+}
+
+impl<T: Clone> OneOrMany<T> {
+    /// Get the first item in the list.
+    pub fn first(&self) -> T {
+        self.first.clone()
+    }
+
+    /// Get the rest of the items in the list (excluding the first one).
+    pub fn rest(&self) -> Vec<T> {
+        self.rest.clone()
+    }
+
+    /// Get all items in the list (joins the first with the rest).
+    pub fn all(&self) -> Vec<T> {
+        let mut all = vec![self.first.clone()];
+        all.extend(self.rest.clone());
+        all
+    }
+}
+
+/// Create a OneOrMany object with a single item.
+impl<T> From<T> for OneOrMany<T> {
+    fn from(item: T) -> Self {
+        OneOrMany {
+            first: item,
+            rest: vec![],
+        }
+    }
+}
+
+/// Create a OneOrMany object with a list of items.
+impl<T> From<Vec<T>> for OneOrMany<T> {
+    fn from(items: Vec<T>) -> Self {
+        let mut iter = items.into_iter();
+        OneOrMany {
+            first: match iter.next() {
+                Some(item) => item,
+                None => panic!("Cannot create OneOrMany with an empty vector."),
```

Comment:
I know that this will seem bad but here's the reason why I did this:

I have to implement `Embeddable` trait for a vector of items that implement `Embeddable`, ie:
```
impl<T: Embeddable> Embeddable for Vec<T> {
      type Error = <SOMETHING>;

     fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error> {
        let items = self
            .iter()
            .map(|item| item.embeddable())
            .collect::<Result<Vec<_>, _>>()?; // <-- T::Error raised here maybe

        Ok(OneOrMany::try_from(items)) // <-- OneToMany empty list error raised here maybe    }
}
```
(see line 209). This is necessary because a big use case is having multiple strings to embed per object. In the `embeddable` function implementation, there are 2 possible error types:
1. `T::Error`
2. Some error which is raised if you try to create a `OneToManyError` with a empty vector.

So what do we set `type Error` to? Obviously, we can't convert `T::Error` to `OneToManyError` and vise versa.
**Option 1:** Create an error type that wraps the 2 error types together, like:
```
#[derive(Debug, thiserror::Error)]
pub enum SomeError {
    #[error("OneToManyError: Can't create OneToMany with empty vector")]
    OneToManyError(()),
    #[error("Error: {}")]
    Error(String),
}
```
this isn't the best IMO because I would have to convert `T::Error` to a string, in essence losing `T`'s actual error type:
```
     let items = self
            .iter()
            .map(|item| item.embeddable())
            .collect::<Result<Vec<_>, _>>().map(|e| SomeError::Error(e.to_string()))?;
```
**Option 2:** Use `T::Error` only
* panic if we attempt to create `OneToMany` with an empty vector. Now, the function has a single error type `T::Error` and we can return it.

Reply:
You could change the `EmbeddableError` to this:
```rust
#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct EmbeddableError(#[from] Box<dyn std::error::Error + Send + Sync>);
```
And you can then convert any error into it, e.g.
```rust
impl Embeddable for serde_json::Value {
    type Error = EmbeddableError;

    fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error> {
        Ok(OneOrMany::one(
            serde_json::to_string(self).map_err(|err| EmbeddableError(Box::new(err)))?,  // <-------- THIS
        ))
    }
}
```

Once that is done, then `OneOrMany::many` can return a `Result`, for which the error can be converted into an `EmbeddableError`